wraps/core/src/api/transaction.rs

```diff
@@ -1,155 +1,171 @@
-        TransactionRequest, H256, U256, Eip1559TransactionRequest,
+        transaction::eip2718::TypedTransaction, Address, Bytes, Eip1559TransactionRequest,
+        TransactionRequest, H256, U256,
     },
-    utils::{serialize},
+    utils::serialize,
 };
-use ethers_core::types::{BlockId, Chain};
 use ethers_providers::ProviderError;
-// use polywrap_wasm_rs::wrap_debug_log;
 
-use crate::error::WrapperError;
-use crate::provider::{PolywrapProvider};
-use crate::signer::PolywrapSigner;
-use crate::polywrap_provider::sync_provider::SyncProvider;
 use crate::mapping::EthersTxOptions;
 
-use crate::api::abi::{tokenize_values, encode_function};
+pub fn create_transaction(
+    address: Option<Address>,
+    data: Bytes,
+    options: &EthersTxOptions,
+) -> TypedTransaction {
+    if options.gas_price.is_some() {
+        return TransactionRequest {
+            to: address.map(Into::into),
+            data: Some(data),
+            gas: options.gas_limit,
+            gas_price: options.gas_price,
+            value: options.value,
+            nonce: options.nonce,
+            ..Default::default()
+        }
+        .into();
+    }
+    Eip1559TransactionRequest {
+        to: address.map(Into::into),
+        data: Some(data),
+        gas: options.gas_limit,
+        max_fee_per_gas: options.max_fee_per_gas,
+        max_priority_fee_per_gas: options.max_priority_fee_per_gas,
+        value: options.value,
+        nonce: options.nonce,
+        ..Default::default()
+    }
+    .into()
+}
 
-pub fn send_transaction(provider: &PolywrapProvider, signer: &PolywrapSigner, tx: &mut TypedTransaction) -> H256 {
-    fill_transaction_sync(provider, signer, tx, None).unwrap();
+pub fn send_transaction(
+    provider: &WrapProvider,
+    signer: &WrapSigner,
+    tx: &mut TypedTransaction,
+) -> H256 {
+    fill_transaction(provider, signer, tx, None).unwrap();
     let rlp = serialize(tx);
-    let tx_hash: H256 = provider.request_sync("eth_sendTransaction", [rlp]).unwrap();
+    let tx_hash: H256 = provider.request("eth_sendTransaction", [rlp]).unwrap();
     tx_hash
 }
 
 pub fn create_deploy_contract_transaction(
     abi: &Abi,
     bytecode: Bytes,
     values: &Vec<String>,
-    options: &EthersTxOptions
+    options: &EthersTxOptions,
 ) -> Result<TypedTransaction, WrapperError> {
     let data: Bytes = match (abi.constructor(), values.is_empty()) {
         (None, false) => {
             let error = "Constructor not found in contract ABI".to_string();
             return Err(WrapperError::ContractError(error));
-        },
+        }
         (None, true) => bytecode.clone(),
         (Some(constructor), _) => {
-            let tokens: Vec<Token> = tokenize_values(&values, &constructor.inputs);
-            constructor.encode_input(bytecode.to_vec(), &tokens)?.into()
-        },
+            let tokens: Vec<Token> = ethers_utils::tokenize_values(&values, &constructor.inputs);
+            //TODO (cbrzn): Remove unwrap
```

Comment:
Could you make this TODO statement clearer so it's obvious what needs to be done? I don't quite understand what you mean here

Reply:
woops, nice catch